src/core/CCList.ml

```diff
@@ -745,6 +745,34 @@ let sorted_merge ~cmp l1 l2 =
     List.length (sorted_merge ~cmp:CCInt.compare l1 l2) = List.length l1 + List.length l2)
 *)
 
+let sorted_diff ~cmp l1 l2 =
+  let rec recurse cmp acc l1 l2 = match l1,l2 with
+    | [], _ -> List.rev acc
+    | _, [] -> List.rev_append acc l1
+    | x1::l1', x2::l2' ->
+      let c = cmp x1 x2 in
+      if c < 0 then recurse cmp (x1::acc) l1' l2
+      else if c > 0 then recurse cmp acc l1 l2'
+      else recurse cmp acc l1' l2'
```

Comment:
```suggestion
      else recurse cmp acc l1' l2
```

if there are several consecutive occurrences of `x1` in `l1` but only one in `l2`, correct?

Reply:
This is intentional---otherwise it would be wrong for multisets. The idea is that `sorted_diff (sorted_merge l1 l2) l2 = l1`. Actually, maybe this should be added to the quickchecking...

Reply:
For my applications it actually does not matter, so I am not strongly against the other behavior.

Reply:
Actually, I am not sure if the other behavior is very useful. If the elements are unique in individual lists, then `sorted_diff` and `sorted_merge_unique` will handle the lists correctly. If the elements are not unique (lists as multisets), then `sorted_diff` and `sorted_merge` will handle them correctly. In either case, the current code seems to work well, and thus I prefer keeping it.

Reply:
My point is, what if you use the sorted list as a multiset and want to remove only one element, not all its occurrences? :)

Reply:
Could you possibly give an example? This thread is for `sorted_diff`, and I felt `sorted_diff` is already doing exactly what you specify. As for `sorted_remove`, it has been adjusted.

Reply:
say with this test:

```ocaml
(*$=
  [1;3] (sorted_diff [1;2;2;3] [2])
*)
```

Reply:
I feel for multisets, it might make more sense to have `{a,a,a,a,b,b} \ {a,a,b} = {a,a,b}`, governed by the formula
```
multiplicity(x\y)(a) = max(0, multiplicity(x)(a) - multiplicity(y)(a))
```
Currently, `sorted_merge l [x]` is the same as `sorted_insert x l`, and for consistency maybe `sorted_diff l [x]` should be the same as `sorted_remove x l`, removing only one element from `l`?

PS: Currently I am not using multisets so none of these would affect me.

Reply:
I agree on the laws for `merge l [x] = insert x l` and the symmetric, but that's also the point: there's already some arguments to specify whether insert can produce duplicates or not, and similarly delete should be able to remove all elements or only one. The default value of the optional argument to control that should be the one that respects the laws.

Reply:
I am sorry but I am still confused. Perhaps you missed the changes to `sorted_remove` which added the optional argument `all` with the default value `false`? By default, `sorted_remove` will remove only one occurrence. If this is different from your impression, I encourage you to check https://github.com/c-cube/ocaml-containers/pull/378/commits/6d2dc4ccf44628a5453ff8a79c677c1e01c9802e which changed the behavior of `sorted_remove`.

One of the following must be false according to your design, but I am not sure which one.

1. `[1;3] = diff [1;2;2;3] [2]` (this is NOT satisfied by the current code)
2. `remove 2 [1;2;2;3] = diff [1;2;2;3] [2]` (this is satisfied by the current code)
3. `[1;2;3] = remove 2 [1;2;2;3]` (this is also satisfied by the current code)

Reply:
(I realized my commit messages did not explain the changes. Sorry about that...)

Reply:
I'm saying there's a `?all:bool` argument to remove that makes 1. true by default, but when set to false makes 3. true. Is that what you did? I'm having a hard time following both commits and discussions, sorry. It might be already what the code does.

Reply:
Current (after https://github.com/c-cube/ocaml-containers/commit/6d2dc4ccf44628a5453ff8a79c677c1e01c9802e) `?all:bool` is `false` by default, which I thought was more consistent with the rest of the library, and my initial choice when drafting this PR seems to be suboptimal. Anyways, I chose the current (default) behaviors based on the following equations for lists as multisets:
```
diff (merge l1 l2) l2 = l1
remove x (insert x l) = l
diff l [x] = remove x l
```
The second equation will force the default value of `?all:bool` to be `false`.

Reply:
A last thing, sorry :sweat_smile: . There's `sorted_merge_uniq` which explicitly targets sets (removes duplicates). I suppose the pendant of that would be `sorted_diff_uniq` (?!) which makes sure to remove all occurrences of common elements?

Reply:
I have to think about it... what are the equations `sorted_diff_uniq` should satisfy?